### PR TITLE
make a separate getShiroBundle function in AirpalApplicationBase

### PR DESCRIPTION
**Pull request by ttyao**
Make a separate getShiroBundle() function so that the subclasses can override just that part without duplicate other bundles' code.

Also make a protected variable injector for easy access to it in subclasses as well.

### CODE REVIEW

**Comment by ttyao** [comment]
@krishnap @andykram

**Comment by ttyao** [comment]
@andykram updated the PR. Added two functions `public Collection<ConfiguredBundle<T>> getConfiguredBundles()` and `public Collection<Bundle> getBundles()` because Bundle and ConfiguredBundle<T> interfaces are not compatible.

File: src/main/java/com/airbnb/airpal/AirpalApplicationBase.java [hunk 1]
```diff
@@ -38,43 +42,50 @@
...
     }
 
     public abstract Iterable<AbstractModule> getModules(T config, Environment environment);
 
+    public Collection<ConfiguredBundle<T>> getConfiguredBundles()
+    {
+        return new ArrayList<>();
+    }
+
+    public Collection<Bundle> getBundles()
```
Comment (andykram) [inline src/main/java/com/airbnb/airpal/AirpalApplicationBase.java]: Make these `Iterable` of `Bundle` and `ConfiguredBundle`

File: src/main/java/com/airbnb/airpal/AirpalApplicationBase.java [hunk 2]
```diff
@@ -38,43 +42,50 @@
...
+    {
+        return new ArrayList<>(Arrays.asList(
```
Comment (andykram) [inline src/main/java/com/airbnb/airpal/AirpalApplicationBase.java]: `return ImmutableList.of(...)`

File: src/main/java/com/airbnb/airpal/AirpalApplicationBase.java [hunk 3]
```diff
@@ -38,43 +42,50 @@
...
+        for (Bundle bundle : getBundles()) {
+            bootstrap.addBundle(bundle);
+        }
     }
 
     public abstract Iterable<AbstractModule> getModules(T config, Environment environment);
 
+    public Collection<ConfiguredBundle<T>> getConfiguredBundles()
+    {
+        return new ArrayList<>();
```
Comment (andykram) [inline src/main/java/com/airbnb/airpal/AirpalApplicationBase.java]: `return Collections.emptyList();`

File: src/main/java/com/airbnb/airpal/AirpalApplication.java [hunk 4]
```diff
@@ -19,6 +23,20 @@
                              new AirpalModule(config, environment));
     }
 
+    @Override
+    public Collection<ConfiguredBundle<AirpalConfiguration>> getConfiguredBundles()
+    {
+        Collection<ConfiguredBundle<AirpalConfiguration>> bundles = super.getConfiguredBundles();
```
Comment (andykram) [inline src/main/java/com/airbnb/airpal/AirpalApplication.java]: Use `ImmutableList.Builder` to construct a new collection.

**Pull request merged**